sstables/index_reader.hh

```diff
@@ -774,6 +774,9 @@ public:
 
     // Advance index_reader bounds to the bounds of the supplied range
     future<> advance_to(const dht::partition_range& range) {
+        if (eof()) {
+            return make_ready_future<>();
+        }
```

Comment:
Why not put it only in `advance_to(index_bound& bound, dht::ring_position_view pos)`? Seems better to put it deeper, so that fewer assumptions leak to callers.

Reply:
Putting it deeper breaks a lot of tests. Maybe it breaks reverse reads? I didn't investigate to be honest.

Reply:
Let me give this another try. Maybe I put it in the wrong `advance_to()` overload.

Reply:
`advance_to(index_bound& bound, dht::ring_position_view pos)` is also called when advancing the end position and `eof()` is based on the position of `_lower_bound`.

Reply:
This can lead to short-circuiting of a request to advance the end bound, when the latter can still be safely advanced. Maybe we should change `eof()` to take a `bound` and check `eof()` from the aspect of said bound.

Reply:
This seems to solve it. I will send a follow-up.

Reply:
Follow-up PR: https://github.com/scylladb/scylla/pull/10531